python/taichi/math/mathimpl.py

```diff
@@ -469,71 +469,111 @@ def mod(x, y):
-        >>>     v = vec2(1, 0)
-        >>>     print(rotate2d(v, radians(30)))
-        [0.866025, 0.500000]
+        >>> import math
+        >>> ti.Matrix.translate(1, 2, 3)
+        [[ 1 0 0 1]
+         [ 0 1 0 2]
+         [ 0 0 1 3]
+         [ 0 0 0 1]]
     """
-    ca, sa = ti.cos(ang), ti.sin(ang)
-    x, y = p
-    return vec2(x * ca - p.y * sa, x * sa + y * ca)
+    return mat4([[1., 0., 0., dx], [0., 1., 0., dy], [0., 0., 1., dz],
+                 [0., 0., 0., 1.]])
 
 
 @ti.func
-def rotate3d(p, axis, ang):
-    """Rotates a vector in 3d space, given an axis and angle of rotation.
-
-    The vector `axis` should be a unit vector.
-
-    See "https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula"
+def scale(sx, sy, sz):
+    """Constructs a scale Matrix with shape (4, 4).
 
     Args:
-        p (:class:`~taichi.math.vec3`): The 3d vector to rotate.
-        axis (:class:`~taichi.math.vec3`): Axis of rotation.
-        ang (float): Angle of rotation, in radians.
+        sx (float): scale x.
+        sy (float): scale y.
+        sz (float): scale z.
+
+    Returns:
+        :class:`~taichi.math.mat4`: scale matrix.
 
     Example::
 
-        >>> from taichi.math import *
-        >>> @ti.kernel
-        >>> def test():
-        >>>     v = vec3(1, 0, 0)
-        >>>     axis = normalize(vec3(1, 1, 1))
-        >>>     print(rotate3d(v, axis, radians(30)))
-        [0.910684, 0.333333, -0.244017]
+        >>> import math
+        >>> ti.Matrix.scale(1, 2, 3)
+        [[ 1 0 0 0]
+         [ 0 2 0 0]
+         [ 0 0 3 0]
+         [ 0 0 0 1]]
+    """
+    return mat4([[sx, 0., 0., 0.], [0., sy, 0., 0.], [0., 0., sz, 0.],
+                 [0., 0., 0., 1.]])
+
+
+@ti.func
+def rot3d_by_axis(ang, axis):
+    """rotate the matrix by an angle with the vector as the rotation axis
+
+    Args:
+        ang (float): angle in radians unit
+        axis (vec3): rotation axis
 
     Returns:
-        :class:`~taichi.math.vec3`: The vector after rotation.
+        :class:`~taichi.math.mat4`: rotation matrix
     """
-    ca, sa = ti.cos(ang), ti.sin(ang)
-    return mix(dot(p, axis) * axis, p, ca) + cross(axis, p) * sa
+    c = ti.cos(ang)
+    s = ti.sin(ang)
+
+    axis = normalize(axis)
+    temp = (1 - c) * axis
+    return mat4([[
+        c + temp[0] * axis[0], temp[0] * axis[1] + s * axis[2],
+        temp[0] * axis[2] - s * axis[1], 0.
+    ],
+                 [
+                     temp[1] * axis[0] - s * axis[2], c + temp[1] * axis[1],
+                     temp[1] * axis[2] + s * axis[0], 0.
+                 ],
+                 [
+                     temp[2] * axis[0] + s * axis[1],
+                     temp[2] * axis[1] - s * axis[0], c + temp[2] * axis[2], 0.
+                 ], [0., 0., 0., 1.]])
 
 
 @ti.func
-def eye(n: ti.template()):
-    """Returns the nxn identity matrix.
+def rot3d_yaw_pitch_roll(yaw, pitch, roll):
+    """Creates a 3D 4 * 4 homogeneous rotation matrix from euler angles(Y * X * Z).
```

Comment:
```suggestion
    """Returns a 4x4 homogeneous rotation matrix representing the 3d rotation with Euler angles (yaw, pitch, roll).
```

Reply:
we should also explain what are the axes of the three rotations and in which order they are applied.

Reply:
I think we should indicate how we rotate with what order and what axis.

```py
Returns a 4x4 homogeneous rotation matrix representing
the 3d rotation with Euler angles (rotate with Y axis fitst, X axis second, Z axis third).
```

Reply:
This should be fine.